### PR TITLE
Add verbosity to the command line

### PR DESCRIPTION
**Pull request by nitro-marky**
Add a verbosity flag if you want to print all the information. 

It's can be run like this `npm run local list --  --verbose`

Without the verbose flag 

`  { id: 'process-1', version: 2, status: 'Enabled' },
  { id: 'process-1', version: 1, status: 'Enabled' },
  { id: 'test-program-propose', version: 1, status: 'Enabled' },
  { id: 'process-2', version: 2, status: 'Enabled' },
  { id: 'process-2', version: 1, status: 'Enabled' },`

with verbose flag

`id: 'process-1',
    version: 2,
    status: 'Enabled',
    program: [
      {
        restriction: { senderHasInputRole: { index: 0, roleKey: 'Supplier' } }
      }
    ]
  },
  {
    id: 'process-1',
    version: 1,
    status: 'Enabled',
    program: [
      {
        restriction: { senderHasInputRole: { index: 0, roleKey: 'Supplier' } }
      }
    ]
  },
  {
    id: 'test-program-propose',
    version: 1,
    status: 'Enabled',
    program: [
      { restriction: { fixedNumberOfInputs: { numInputs: 2 } } },
      { restriction: { fixedNumberOfOutputs: { numOutputs: 3 } } },
      { op: 'And' },
      {
        restriction: {
          fixedInputMetadataValue: {
            index: 0,
            metadataKey: '0x74797065',
            metadataValue: { literal: '0x44454d414e44' }
          }
        }
      },
      { op: 'And' },`

----------------------------------------------------------------------------------------------------------------------------

Todo in another PR:

- Unit testing on the verbose additions

### CODE REVIEW

**Review by dblane-digicatapult — COMMENTED** [review]
Is this the first of several PRs for this L3-138?

**Comment by nitro-marky** [comment]
> Is this the first of several PRs for this L3-138?

@n3op2 said this is a good place to start, is there anything in particular you want adding to this?

**Comment by nitro-marky** [comment]
> Is this the first of several PRs for this L3-138?

I see what you mean, add one for create

File: src/index.ts [hunk 1]
```diff
@@ -69,14 +69,14 @@ program
 
       if (raw) {
         dir(processes, { depth: null })
-      } else {
+      } else{
```
Comment (mattdean-digicatapult) [inline src/index.ts:72]: This is concerning there's a whitespace issue here that the linter should have caught but didn't. @n3op2 can you have a look and see if you can work out what's wrong in the linter config

File: src/index.ts [hunk 2]
```diff
@@ -57,7 +58,6 @@ program
     try {
       const res: Process.RawPayload[] = await getAll(mapOptions(options))
       const { raw, active, disabled } = options
-
       let processes: Process.RawPayload[]
```
Comment (mattdean-digicatapult) [inline src/index.ts:61]: I think there's some refactoring that can be done here to improve readability and testability. Basically from ln 61 to ln 84 we filter or manipulate `res` into something that we then pass to `dir`. My recommendation would be to refactor the logic of that manipulation into a pure helper function that we can then unit test. The code here would then just call `getAll`, transform the output, and call `dir` on that output. Does that make sense?

Comment (nitro-marky) [reply src/index.ts:61]: 👍

File: src/lib/process/index.ts [hunk 3]
```diff
@@ -36,28 +36,60 @@ export const loadProcesses = async ({
...
+  if (options.active) {
+    processes = res.filter(({ status }) => status === 'Enabled')
+  } else if (options.disabled) {
+    processes = res.filter(({ status }) => status === 'Disabled')
+  } else {
+    processes = res
+  }
+
+  if (options.raw) {
+    return [processes, { depth: null }]
```
Comment (mattdean-digicatapult) [inline src/lib/process/index.ts:70]: Probably better off just returning the `processes` and not the `{ depth: null }`. The latter parameter is the concern of `dir` which is called elsewhere.

File: src/lib/process/index.ts [hunk 4]
```diff
@@ -36,28 +36,60 @@ export const loadProcesses = async ({
...
+  } else{
+    return(
```
Comment (mattdean-digicatapult) [inline src/lib/process/index.ts:72]: again I'd update the return here to just return the output from the `map`

File: src/lib/process/index.ts [hunk 5]
```diff
@@ -36,28 +36,60 @@ export const loadProcesses = async ({
...
   for (let i = 0; i < processes.length; i++) {
     const { name, version, program } = processes[i]
-    res[name] = await createProcess(name, version, program, dryRun, options)
+    res[name] = await createProcess(name, version, program, dryRun, options, verbose)
   }
 
   return res
 }
 
+export const listProcess = async (
```
Comment (mattdean-digicatapult) [inline src/lib/process/index.ts:57]: This is being picky but the name of this function here feels incorrect. We're not listing processes. We're transforming the processes for output

File: src/lib/process/api.ts [hunk 6]
```diff
@@ -40,7 +41,7 @@ export const createProcessTransaction = async (
             id: data[0].toHuman(),
             version: data[1].toNumber(),
             status: 'Enabled',
-            program,
+            ...verbose ? program : {},
```
Comment (mattdean-digicatapult) [inline src/lib/process/api.ts:44]: Is there a reason we're filtering here? The concern of this function is performing the transaction, yet we're conflating that with adjusting cli output?

File: src/index.ts [hunk 7]
```diff
@@ -106,9 +84,9 @@ program
       options: ${b(JSON.stringify(options))}
       program: ${b(data)}
     `)
-    const { dryRun } = options
+    const { dryRun, verbose } = options
     try {
-      const res: Process.Response = await loadProcesses({ data, dryRun, options: mapOptions(options) })
+      const res: Process.Response = await loadProcesses({ data, dryRun, options: mapOptions(options), verbose })
```
Comment (n3op2) [inline src/index.ts:89]: Maybe we could include verbose in mapOptions? less places to update, would need to go through to confirm tho as it looks like polkadot related only. https://github.com/digicatapult/dscp-process-management/blob/83f0e6d45775ee86202367c53a5194390b094d9c/src/index.ts#L28C1-L32C3

```
const mapOptions = (options: Process.CLIOptions): Polkadot.Options => ({
  API_HOST: options.host,
  API_PORT: parseInt(options.port),
  USER_URI: options.user,
})
```

File: src/lib/process/api.ts [hunk 8]
```diff
@@ -43,6 +44,11 @@ export const createProcessTransaction = async (
             program,
           }
 
+          if(verbose == false)
+          {
+            delete newProcess.program
```
Comment (n3op2) [inline src/lib/process/api.ts:49]: purely FYI: https://dev.to/hardiquedasore/avoid-delete-keyword-in-javascript-to-remove-property-32mf there are a number of opinions but for now if it works it works

**Review by n3op2 — COMMENTED** [review]
looks like we are getting there

File: src/index.ts (hunk 7 above)
Comment (nitro-marky) [reply src/index.ts:89]: I'll leave it for now, and focus on getting the tests done so we can merge.

File: tests/integration/command-functions.test.ts [hunk 9]
```diff
@@ -113,6 +117,30 @@ describe('Process creation and deletion, listing', () => {
         version: bumpedVersion,
         status: 'Disabled',
       })
+
+      /*
```
Comment (n3op2) [inline tests/integration/command-functions.test.ts:121]: was not aware that this will get pushed over this branch. could you please remove comments. if this will be included as part of another PR?

File: tests/integration/command-functions.test.ts [hunk 10]
```diff
@@ -138,14 +166,16 @@ describe('Process creation and deletion, listing', () => {
       })
     })
 
-    it('returns a list of raw processes', async () => {
+    // it('returns a list of raw processes when with --verbose flag', () => {
```
Comment (n3op2) [inline tests/integration/command-functions.test.ts:169]: same here

**Review by n3op2 — CHANGES_REQUESTED** [review]

**Review by mattdean-digicatapult — COMMENTED** [review]
once @n3op2 is happy I'm happy

**Review by n3op2 — APPROVED** [review]
